File: gma-next-gen-sdk-android-examples-main/java/nextgenexample/app/src/main/java/com/example/nextgenexample/inlinebanner/RecyclerViewAdapter.java
package com.example.nextgenexample.inlinebanner;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;
import com.example.nextgenexample.Constant;
import com.example.nextgenexample.databinding.InlineBannerAdBinding;
import com.example.nextgenexample.databinding.RecyclerViewMenuItemBinding;
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAd;
import java.util.List;

/**
 * The [RecyclerViewAdapter] class.
 * The adapter provides access to the items in the [MenuItemViewHolder] or the [BannerAdHolder].
 */
public final class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

  // A menu item view type.
  private static final int MENU_ITEM_VIEW_TYPE = 0;
  // The banner ad view type.
  private static final int BANNER_AD_VIEW_TYPE = 1;
  // An Activity.
  private final FragmentActivity activity;
  // The recyclerViewItems list contains only [InlineMenuItem] and [BannerAd] types.
  private final List<Object> recyclerViewItems;

  private RecyclerViewAdapter(FragmentActivity activity, List<Object> recyclerViewItems) {
    this.activity = activity;
    this.recyclerViewItems = recyclerViewItems;
  }

  /** Return a new instance of [RecyclerViewAdapter]. */
  public static RecyclerViewAdapter newInstance(FragmentActivity activity, List<Object> recyclerViewItems) {
    return new RecyclerViewAdapter(activity, recyclerViewItems);
  }

  private static class MenuItemViewHolder extends RecyclerView.ViewHolder {
    private final RecyclerViewMenuItemBinding binding;

    MenuItemViewHolder(RecyclerViewMenuItemBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }

    TextView getMenuItemName() {
      return binding.menuItemName;
    }
    TextView getMenuItemDescription() {
      return binding.menuItemDescription;
    }
    TextView getMenuItemPrice() {
      return binding.menuItemPrice;
    }
    TextView getMenuItemCategory() {
      return binding.menuItemCategory;
    }
    ImageView getMenuItemImage() {
      return binding.menuItemImage;
    }
  }

  private static class BannerAdHolder extends RecyclerView.ViewHolder {
    private final InlineBannerAdBinding binding;

    BannerAdHolder(InlineBannerAdBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }

    public CardView getBannerViewContainer() {
      return binding.bannerViewContainer;
    }
  }

  @Override
  public int getItemCount() {
    return recyclerViewItems.size();
  }

  /**
   * Determines the view type for the given position.
   */
  public int getItemViewType(int position) {
    if (recyclerViewItems.get(position) instanceof BannerItem) {
      return BANNER_AD_VIEW_TYPE;
    } else {
      return MENU_ITEM_VIEW_TYPE;
    }
  }

  /**
   * Creates a new view for a menu item view or a banner ad view based on the viewType. This method
   * is invoked by the layout manager.
   */
  @NonNull
  @Override
  public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
    LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());

    if (viewType == MENU_ITEM_VIEW_TYPE) {
      RecyclerViewMenuItemBinding binding =
          RecyclerViewMenuItemBinding.inflate(inflater, viewGroup, false);
      return new MenuItemViewHolder(binding);
    } else {
      InlineBannerAdBinding binding =
          InlineBannerAdBinding.inflate(inflater, viewGroup, false);
      return new BannerAdHolder(binding);
    }
  }

  /**
   * Replaces the content in the views that make up the menu item view and the banner ad view. This
   * method is invoked by the layout manager.
   */
  @Override
  public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
    Object item = recyclerViewItems.get(position);

    if (holder instanceof MenuItemViewHolder menuItemHolder
        && item instanceof FoodMenuItem menuItem) {
      // Get the menu item image resource ID.
      String imageName = menuItem.imageName();
      int imageResID =
          activity.getResources().getIdentifier(imageName, "drawable", activity.getPackageName());

      // Add the menu item details to the menu item view.
      menuItemHolder.getMenuItemImage().setImageResource(imageResID);
      menuItemHolder.getMenuItemName().setText(menuItem.name());
      menuItemHolder.getMenuItemPrice().setText(menuItem.price());
      menuItemHolder.getMenuItemCategory().setText(menuItem.category());
      menuItemHolder.getMenuItemDescription().setText(menuItem.description());

    } else if (holder instanceof BannerAdHolder bannerHolder
        && item instanceof BannerItem bannerItem) {
      if (bannerItem.bannerAd != null) {
        // Get the CardView container from the AdViewHolder.
        CardView bannerViewContainer = bannerHolder.getBannerViewContainer();

        // Add the banner ad view to the CardView container.
        BannerAd bannerAd = bannerItem.bannerAd;
        bannerViewContainer.addView(bannerAd.getView(activity));
      }
    } else {
      Log.e(Constant.TAG, "Invalid type at position: " + position);
    }
  }

  @Override
  public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
    if (holder instanceof BannerAdHolder bannerHolder) {
      // Detach the banner ad from the recycled view holder.
      CardView bannerViewContainer = bannerHolder.getBannerViewContainer();
      if (bannerViewContainer != null) {
        bannerViewContainer.removeAllViews();
      }
    }
  }
}

File: gma-next-gen-sdk-android-examples-main/kotlin/nextgenexample/app/src/main/java/com/example/next_gen_example/inlinebanner/RecyclerViewAdapter.kt
package com.example.next_gen_example.inlinebanner

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import com.example.next_gen_example.Constant
import com.example.next_gen_example.databinding.InlineBannerAdBinding
import com.example.next_gen_example.databinding.RecyclerViewMenuItemBinding

class RecyclerViewAdapter(
  private val activity: FragmentActivity,
  private val recyclerViewItems: List<Any>,
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

  private class MenuItemViewHolder(binding: RecyclerViewMenuItemBinding) :
    RecyclerView.ViewHolder(binding.root) {

    val menuItemName: TextView = binding.menuItemName
    val menuItemDescription: TextView = binding.menuItemDescription
    val menuItemPrice: TextView = binding.menuItemPrice
    val menuItemCategory: TextView = binding.menuItemCategory
    val menuItemImage: ImageView = binding.menuItemImage
  }

  private class BannerAdViewHolder(private val binding: InlineBannerAdBinding) :
    RecyclerView.ViewHolder(binding.root) {

    fun getBannerViewContainer(): CardView = binding.bannerViewContainer
  }

  override fun getItemCount(): Int = recyclerViewItems.size

  /** Determines the view type for the given position. */
  override fun getItemViewType(position: Int): Int {
    return if (recyclerViewItems[position] is BannerItem) {
      BANNER_AD_VIEW_TYPE
    } else {
      MENU_ITEM_VIEW_TYPE
    }
  }

  /**
   * Creates a new view for a menu item view or a banner ad view based on the viewType. This method
   * is invoked by the layout manager.
   */
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
    val inflater = LayoutInflater.from(parent.context)
    return if (viewType == MENU_ITEM_VIEW_TYPE) {
      val binding = RecyclerViewMenuItemBinding.inflate(inflater, parent, false)
      MenuItemViewHolder(binding)
    } else {
      val binding = InlineBannerAdBinding.inflate(inflater, parent, false)
      BannerAdViewHolder(binding)
    }
  }

  /**
   * Replaces the content in the views that make up the menu item view and the banner ad view. This
   * method is invoked by the layout manager.
   */
  override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
    when (holder) {
      is MenuItemViewHolder -> {
        val menuItem = recyclerViewItems[position] as? FoodMenuItem
        if (menuItem == null) {
          Log.w(Constant.TAG, "FoodMenuItem not found at position $position.")
          return
        }

        // Get the menu item image resource ID.
        val imageName = menuItem.imageName
        val imageResID =
          activity.resources.getIdentifier(imageName, "drawable", activity.packageName)

        // Add the menu item details to the menu item view.
        holder.menuItemImage.setImageResource(imageResID)
        holder.menuItemName.text = menuItem.name
        holder.menuItemPrice.text = menuItem.price
        holder.menuItemCategory.text = menuItem.category
        holder.menuItemDescription.text = menuItem.description
      }
      is BannerAdViewHolder -> {
        val bannerItem = recyclerViewItems[position] as? BannerItem
        if (bannerItem == null) {
          Log.w(Constant.TAG, "BannerItem not found at position $position.")
          return
        }

        if (bannerItem.isBannerAdInitialized()) {
          // Get the CardView container from the BannerAdViewHolder.
          val bannerViewContainer = holder.getBannerViewContainer()

          val bannerAd = bannerItem.bannerAd
          // Add the banner ad view to the CardView container.
          bannerViewContainer.addView(bannerAd.getView(activity))
        }
      }
    }
  }

  override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
    super.onViewRecycled(holder)
    if (holder is BannerAdViewHolder) {
      // Detach the banner ad from the recycled view holder.
      val bannerViewContainer = holder.getBannerViewContainer()
      bannerViewContainer.removeAllViews()
    }
  }

  private companion object {
    // A menu item view type.
    const val MENU_ITEM_VIEW_TYPE = 0
    // The banner ad view type.
    const val BANNER_AD_VIEW_TYPE = 1
  }
}
